Key Usage, Basic Constraints, Extended Key Usage

dannyallover opened the pull request.

dannyallover commented:
PTAL @sleevi

dannyallover commented:
Note: working on adding some comments + updating once https://github.com/google/fuzzing/pull/65 is merged.

sleevi commented on proto/asn1-pdu/x509_certificate_to_der.cc:
```diff
@@ -43,6 +43,172 @@ DECLARE_ENCODE_FUNCTION(AlgorithmIdentifierSequence) {
+
+DECLARE_ENCODE_FUNCTION(KeyUsage) {
+  uint16_t key_usage = 0x00;
+  auto masks = {val.digital_signature() ? 0x01 : 0x00,
+                val.digital_signature() ? 0x02 : 0x00,
+                val.digital_signature() ? 0x04 : 0x00,
+                val.digital_signature() ? 0x08 : 0x00,
+                val.digital_signature() ? 0x10 : 0x00,
+                val.digital_signature() ? 0x20 : 0x00,
+                val.digital_signature() ? 0x40 : 0x00,
+                val.digital_signature() ? 0x80 : 0x00,
+                val.digital_signature() ? 0x100 : 0x00};
```
BUG: This copy/pastes the same bool check, which probably isn't what you want?

You 'should' be able to get away with something like (and sorry, this is where coding in a comment box is terrible)

```suggestion
constexpr struct {
  using Fn = bool (KeyUsage::*)(void);
  Fn function;
  uint32_t mask;
} kValues[] = {
  {
    { &KeyUsage.digital_signature, 0x01 },
    { &KeyUsage.non_repudiation, 0x02 },
    // ...
   }
};
for (const auto& usage : kValues) {
  if ((val.*(usage.function))()) {
  // alt: if std::invoke(val, usage.function), but that's C++17 and best to avoid for now
    key_usage |= usage.mask;
  }
}
```

I think that's right. In the structure, we take pointers to the function to call and the mask to use. In the usage, we then dereference that function, call it, and if it's set, add the mask.

sleevi commented on proto/asn1-pdu/x509_certificate_to_der.cc:
```diff
@@ -43,6 +43,172 @@ DECLARE_ENCODE_FUNCTION(AlgorithmIdentifierSequence) {
+}
+
+DECLARE_ENCODE_FUNCTION(BasicConstraints) {
+  if (val.ca().val()) {
```
Maybe add a note here that it's omitted for default value

```suggestion
  // RFC 5280, 4.2.1.9
  //    BasicConstraints ::= SEQUENCE {
  //        cA                      BOOLEAN DEFAULT FALSE,
  //        pathLenConstraint       INTEGER (0..MAX) OPTIONAL }
  //
  //  Omit basicConstraints if it's false, as that's a DEFAULT value in a SEQUENCE (see
  // X.690 (2015), 11.5)
  if (val.ca().val()) {
```

sleevi commented on proto/asn1-pdu/x509_certificate.proto:
```diff
@@ -134,13 +134,86 @@ message UniqueIdentifier {
+
+// See RFC 5280, 4.2.1.12.
+message ExtendedKeyUsage {
+  required asn1_universal_types.ObjectIdentifier key_purpose_id = 1;
+  repeated asn1_universal_types.ObjectIdentifier key_purpose_ids = 2;
```
Let's add a comment here as well.

```suggestion
  // ExtendedKeyUsage is SEQUENCE SIZE (1..MAX) OF KeyPurposeId,
  // so always ensure at least one key purpose ID is provided.
  required asn1_universal_types.ObjectIdentifier key_purpose_id = 1;
  repeated asn1_universal_types.ObjectIdentifier key_purpose_ids = 2;
```

dannyallover commented:
PTAL @sleevi

sleevi commented on proto/asn1-pdu/x509_certificate.proto:
```diff
@@ -165,9 +165,12 @@ message Extension {
     BasicConstraints basic_constraints = 7;
     ExtendedKeyUsage extended_key_usage = 8;
+=======
+>>>>>>> master
```
Left-over merge markers to be cleaned up? :)

sleevi commented on proto/asn1-pdu/x509_certificate_to_der.cc:
```diff
@@ -51,30 +51,53 @@ DECLARE_ENCODE_FUNCTION(ExtendedKeyUsage) {
 DECLARE_ENCODE_FUNCTION(KeyUsage) {
+  constexpr struct Mask {
+    using Fn = bool (KeyUsage::*)(void) const;
+    Fn use_mask;
+    uint32_t mask;
```
Sorry, my naming was so bad, I should have done better for a suggestion.

```suggestion
    Fn key_usage_set_fn;
    uint32_t key_usage_value;
```

I'm not sure that's much better, but I'm specifically trying to make lines 93-96 be a bit more descriptive. I think abbreviating "function" to "fn" here should be fine (in the variable name), but I think the goal is, C++ syntax notwithstanding, can you read line 93-96 and make sense of what's going to happen?

dannyallover replied:
I agree with you on all your points! Just wondering here if the name `key_usage_use_value` makes more sense here.

Yes, we are setting the function. But when you invoke the function in the for loop, it feels weird to call `key_usage_set_fn` since we aren't setting anything and rather trying to determine whether we should use the `key_usage_value`.

Let me know if I'm missing something here! :)

sleevi replied:
As the name of the function? That feels a bit weird because the bool doesn't return the value. I had been thinking of the function name in the context of what it returns, and sort of

```
if (foo.key_usage_set()) {
  bar |= foo.key_usage_value
}
```

as the construct we're trying to mirror here. Although I suppose

```
if (foo.is_usage_set()) {
  bar |= foo.usage_value;
}
```

could also work?

dannyallover replied:
Ah, I see what you mean now by the naming `key_usage_set`. That makes sense, and I prefer that.

dannyallover commented:
There were a few bugs where the sequence was not being encoded (when it should be according to the spec). I also added comments where I thought was helpful for a reader. I also addressed the above issue on naming.

I'll have to go over it again later on with a fine tooth comb, but for now I think it looks good.

PTAL @sleevi

dannyallover commented:
PTAL @Dor1s
